File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int prefix[] = new int[nums.length];
        int postfix[] = new int[nums.length];
        int ans[] = new int[nums.length];
        prefix[0] = 1;
        for(int i = 1; i < nums.length; i++){
            prefix[i] = prefix[i -1] * nums[i-1]; 
        }
        int n = nums.length-1;
        postfix[n] = 1;
        for(int i = nums.length-2; i >= 0; i--){
            postfix[i] = postfix[i+1] * nums[i+1];
        }
        for(int i  = 0 ; i < nums.length; i++){
            ans[i] = prefix[i] * postfix[i];
        }        
        return ans;
    }
}
/*int ans[] = new int[nums.length];
        for(int i = 0; i < nums.length; i++){
            int num = nums[i];
            int mul = 1;
            for(int j = 0; j < nums.length; j++){
                if(nums[j] != num){
                    mul *= nums[j];
                }
            }
            ans[i] = mul;
        }
        return ans;*/

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        # 1, 2, 6, 12 --- 24,24,12,4 --- 24, 12, 
        res = [1] * len(nums)
        prefix = 1
        for i in range(len(nums)):
            res[i] = prefix
            prefix *= nums[i]
        postfix = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= postfix
            postfix *= nums[i]
        return res